### PR TITLE
feat(rpc): Hide old banned mns by default

### PR DESCRIPTION
**Pull request by ogabrielides**
## What was done?

It was requested by service desk to hide old banned masternodes when calling rpc `masternodelist`.
The period from which a masternode is considered old banned is more than a `SuperblockCycle`.

| Network        | SuperblockCycle           |
| ------------- |:-------------:|
| Mainnet      | 16616 |
| Testnet    | 24      |
| Devnet | 24     |
| Regtest | 10      |

The new mode `recent` was added to in order to hide old banned masternodes.
Note: If the mode `recent` is used, then the reply mode is `JSON` (can be additionally filtered)

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

**Review by thephez — COMMENTED** [review]
Concept ACK
This seems like a reasonable idea. Curious how 1 month was chosen - was this based on anything specific? And not for this PR, but it would be nice to be able to filter out the "old" banned MNs from the Masternode List tab in Qt also.

**Comment by ogabrielides** [comment]
> Concept ACK This seems like a reasonable idea. Curious how 1 month was chosen - was this based on anything specific? And not for this PR, but it would be nice to be able to filter out the "old" banned MNs from the Masternode List tab in Qt also.

The input was given empirically by the service desk team
I guess statistically if a masternode is banned since more than a month, it will very likely stay banned.

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
concept NACK

**Comment by ogabrielides** [comment]
@UdjinM6 Inverted the logic. Now the logic is by default show everyone, when calling with `recent` show only enabled + recently banned.

File: src/rpc/masternode.cpp [hunk 1]
```diff
@@ -587,7 +588,16 @@ static UniValue masternodelist(const JSONRPCRequest& request)
         return (int)pindex->nTime;
     };
 
+    bool showAllBannedMns = strMode != "recent";
+    int tipHeight = WITH_LOCK(cs_main, return ::ChainActive().Tip()->nHeight;);
+    constexpr int monthBlockLength = 24 * 24 * 30;
     mnList.ForEachMN(false, [&](auto& dmn) {
+        if (!showAllBannedMns && mnList.IsMNPoSeBanned(dmn)) {
```
Comment (UdjinM6) [inline src/rpc/masternode.cpp:595]: `showAllBannedMns` is `!= recent` and then we use `!showAllBannedMns` so it's double-negative which is a bit confusing. Could maybe refactor it to `showRecentMnsOnly` or drop the variable and simply use
```suggestion
        if (strMode == "recent" && mnList.IsMNPoSeBanned(dmn)) {
```

File: src/rpc/masternode.cpp [hunk 2]
```diff
@@ -587,7 +588,16 @@ static UniValue masternodelist(const JSONRPCRequest& request)
         return (int)pindex->nTime;
     };
 
+    bool showAllBannedMns = strMode != "recent";
+    int tipHeight = WITH_LOCK(cs_main, return ::ChainActive().Tip()->nHeight;);
+    constexpr int monthBlockLength = 24 * 24 * 30;
```
Comment (UdjinM6) [inline src/rpc/masternode.cpp:593]: nit: s/Length/Count/ maybe?

also, could probably use `Params().GetConsensus().nSuperblockCycle` instead - would make it much easier to write tests for this mode cause regtest cycles are much smaller (10 blocks). Though I'm not sure if service desk guys are going to be happy about it cause testnet/devnet cycles are also pretty small (24 blocks).

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

**Comment by ogabrielides** [comment]
@UdjinM6 Applied suggestions
@thephez the period was changed to `SuperblockCycle`

**Review by UdjinM6 — APPROVED** [review]
utACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK for squash merge, sure

**Pull request merged**